### PR TITLE
Change error to warning if state_dict is empty in `load_from_checkpoint`

### LINKED ISSUE #18240: Support loading checkpoints with empty `state_dict`
### Description & Motivation

Currently loading checkpoints with empty `state_dict` will get `ValueError(f"The state dict in {checkpoint_path!r} contains no parameters.")`:

https://github.com/Lightning-AI/lightning/blob/c8062021dac9d47fde5cc50338c55c5dbb15e200/src/lightning/pytorch/core/saving.py#L98-L99

We may not need the check anymore since the `cpu` will be the device for the returned module when the `state_dict` is empty:

https://github.com/Lightning-AI/lightning/blob/c8062021dac9d47fde5cc50338c55c5dbb15e200/src/lightning/pytorch/core/saving.py#L98-L103

Could we remove the check above to allow loading checkpoints with empty `state_dict`?

### Pitch

Remove https://github.com/Lightning-AI/lightning/blob/c8062021dac9d47fde5cc50338c55c5dbb15e200/src/lightning/pytorch/core/saving.py#L98-L99 to allow loading checkpoints with empty `state_dict`.

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @awaelchli

### PR DESCRIPTION
**Pull request by ryan597**
## What does this PR do?

`load_from_checkpoint` raises an error if the state_dict in the checkpoint is empty, change this to a warning and return the created model on cpu with the rest of the checkpoint parameters.

Add a test that models can be loaded from checkpoint with empty `state_dict`

Motivation: in #18240 a user has use of loading checkpoints with empty `state_dict` which was possible in previous versions of lightning

Fixes #18240

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: tests/tests_pytorch/core/test_saving.py [hunk 1]
```diff
@@ -103,3 +105,17 @@ def set_extra_state(self, state):
...
     assert model.device.type == "cuda"
+
+
+def test_load_from_checkpoint_warn_on_empty_state_dict(tmp_path):
+    """Test that checkpoints can be loaded with an empty state dict and that the appropriate warning is raised."""
+    create_boring_checkpoint(tmp_path, BoringModel(), accelerator="cpu")
+    # Now edit so the state_dict is empty
+    checkpoint = torch.load(f"{tmp_path}/checkpoint.ckpt")
+    checkpoint["state_dict"] = {}
+    torch.save(checkpoint, f"{tmp_path}/checkpoint.ckpt")
```
Comment (awaelchli) [inline tests/tests_pytorch/core/test_saving.py:116]:
```suggestion
    checkpoint = torch.load(tmp_path / "checkpoint.ckpt")
    checkpoint["state_dict"] = {}
    torch.save(checkpoint, tmp_path / "checkpoint.ckpt")
```

File: tests/tests_pytorch/core/test_saving.py [hunk 2]
```diff
@@ -103,3 +105,17 @@ def set_extra_state(self, state):
...
+
+    with warnings.catch_warnings(record=True) as w:
+        model = BoringModel.load_from_checkpoint(f"{tmp_path}/checkpoint.ckpt", strict=False)
+        assert "contains no parameters" in str(w[-1].message)
+        assert model.device.type == "cpu"
```
Comment (awaelchli) [inline tests/tests_pytorch/core/test_saving.py:121]:
```suggestion
    with pytest.warns(UserWarning, match="contains no parameters"):
        model = BoringModel.load_from_checkpoint(tmp_path / "checkpoint.ckpt", strict=False)
    assert model.device.type == "cpu"
```

**Review by awaelchli — COMMENTED** [review]
Good stuff!

**Review by awaelchli — APPROVED** [review]
Thanks! 🎉
